draft-ietf-taps-interface.md

```diff
@@ -521,18 +522,54 @@ strict requirement, as it signifies a reduction in reliability. The
 default is full checksum coverage without being able to change it, and
 requiring a checksum when receiving.
 
-### Interface Type {#prop-intf-type}
+### Interface Instance or Type {#prop-interface}
 
 Type: Tuple (Enumeration, Preference)
 
-This property specifies which kind of access network interface,
-e.g., WiFi, Ethernet, or LTE, to prefer over others for this Connection, in
-case they are available. In general, Interface Types should be used only with
-the `Prefer` and `Prohibit` preference level. Specifically, using the
-`Require` preference level for Interface Type may limit path selection in a
-way that is detrimental to connectivity. The default is to use the default
-interface configured in the system policy.
-The valid values for the access network interface kinds are implementation specific.
+This property allows the application to select which specific network interfaces
+or categories of interfaces it wants to `Require`, `Prohibit`, or `Prefer`.
+
+If a system supports discovery of specific interface identifiers, such as `en0` or `eth0`
+on Unix-style systems, an implemention should allow using these identifiers to
+define path preferences. Note that marking a specific interface as `Required`
+strictly limits path selection to a single interface, and leads to less flexible and
+resilient connection establishment.
+
+The set of valid interface types is implementation- and system-specific. For example,
+on a mobile device, there may be `Wi-Fi` and `Cellular` interface types available;
+whereas on a desktop computer, there may be `Wi-Fi` and `Wired Ethernet`
+interface types available. Implementations should provide all types that are supported
+on some system to all systems, in order to allow applications to write generic code.
```

Comment:
No, we decided on the call that a registry is definitely not the way to go—I'm saying here that any given implementation likely has a list of types it supports, and it should make all types it knows about available across various platforms to make the code as reusable as possible.

Reply:
Right, that's what I thought, but the text makes it seem like *different* implementations might share these same types. You do mention that these are implementation specific, yet there's no accompanying example like there is for the system-specific types. Maybe we could just add one such example?

Reply:
I think we should be more concrete and more prescriptive for the sake of portability. I'd replace "Implementations should provide all types that are supported on some system to all systems, in order to allow applications to write generic code." with 'For the sake of application portability, all implementations should provide at least the types "Wi-Fi", "Cellular", "Wired Ethernet".'   - these types are pretty generic anyway, what's the harm in always offering them?  I'm afraid that a relaxed statement like the one that's now here is not going to lead to a coherent set across systems (because the point of reference are only other implementations, not this draft!), and so the idea of application portability is lost.

Reply:
I can see that. But what about my counter-argument that saying "Implementations should provide all types that are supported on some system to all systems" isn't ever going to give us enough coherence across implementations that application code is going to really be portable?  What's a better alternative?

Reply:
I'm not sure we need a common interface type system for portability here, because in most case the specific semantics of whether a given application should or should not use a given transport/interface are administrative- or user-level choices, not application developer choices.

In other words, a system would provide a way for an application to ask a user (or, as in Apple's case, provide a system UI to ask the user) which apps should work with which interfaces. This is probably an implementation-level thing though, as it might also tie into other specifics of the platform...

Reply:
I don't understand what this has to do with hard-coding L2-style names or not. Indeed, admin choices will tell us which interfaces can be used, but how can you write an application and port it between systems if you have no common name to use for the WiFi interface?  But as I said, of course I also understand your other argument, that hardcoding L2 names isn't going to be great for all time in the future. I just wanted to raise the concern, I think it would the better compromise to hard-code these names after all, but I'm not going to fight for a paricular choice here. My view is that Tommy should read our discussion and then decide, and let's go for what he picks.

Reply:
> but how can you write an application and port it between systems if you have no common name to use for the WiFi interface?

I guess I'm having a hard time understanding why a (non-network-admin-tool) would want, as part of its application logic, as opposed to in response to some user configuration demand, to define a constraint on interface types. I can think of one such constraint: having two connections, one "essential" and one "higher-bandwidth", and constraining the latter to cheaper connections, but that's really a higher-level policy which (1) requires user input and (2) probably should be expressed in higher-level goals than "wifi or not"...

I think this is a larger discussion (and indeed is kind of the problem PvD is meant to solve)... for this particular issue, though, I think what's here is good. I'm not going to fight awfully hard either, so +1 to making this @tfpauly's problem :)

Reply:
Yes, +1 on this being difficult to understand - I was also just beginning to think along the same lines, that we should abstract from "WiFi" into "cheap" or whatever. Anyway, we agree: this is in @tfpauly 's hands - case closed, problem solved   :-D

Reply:
I read this part similar to @chris-wood, that different implementations would have to provide the same types, or even provide *all* types that any other implementation out there provides. If it means "an implementation that can run on multiple platforms should provide the same types on all these platforms", that's fine with me, but it should be made a little more clear in the text.

I think generally it's good to offer more types and live with the fact that not all types are going to be available on every system an implementation is running on. An implementation will still know what types are actually available on this particular system at run time. If the application asks to "Require" something that isn't there, the implementation will error out early according to Section 3.1 in the Implementation Draft. Maybe this is worth another sentence in the "types" paragraph - a "Require" on a particular interface type will also restrict the selection, maybe unnecessarily.

Imo this kind of interface selection is often about assumptions that an application developer has about a particular type of access network - like "WiFi = cheap" as @mwelzl was saying, and as @britram probably meant by "higher-level goal". I agree with them that for such use-cases we should abstract from specific interface choice and encourage applications to request what they actually want from the network (e.g., cheap), instead of relying on the developer's assumptions, which might be wrong. This is what we are trying to achieve with Intents, too.

Reply:
I can live with a spirit of the text @tfpauly wrote  – But I guess the Interface Type or Instance should only be a corner case anyway. It is implementation specific and I guess it will have surprising results.
What should the system do, if one specifies "prohibit cellular" but it knows that the uplink for the Wifi interface is also cellular?

We should add a warning against using "Wifi" as a representant for "cheap" or "Cellular" as a representant for "expensive" – for this use case, we should use something like the cost preference Intent.

Reply:
Added a comment warning against using interface types as a proxy for metered and unmetered networks.

Reply:
I can not parse `on some system to all systems` in the last paragraph.

I think the whole last paragraph still hard to understand as it introduces a new concept without saying so – how about:

   The set of valid interface types is implementation- and system-specific and can include mnemonic types.
   On a mobile device, for example, there may be the interface types`Wi-Fi` and `Cellular` available; on a desktop computer, there may be `Wi-Fi` and `Wired Ethernet` available.
   Implementations should provide a way to enumerate interface mnemonic types order to allow applications to write generic code.